File: 백준/Silver/2644. 촌수계산/촌수계산.py
import sys

num = 0

def dfs(start, num):
    visited[start] = 1
    #global num
    num += 1

    if y == start:
        result.append(num)
    else:
        for i in graph[start]:
            if not visited[i]:
                dfs(i, num)


n = int(input())
x, y = map(int, input().split())
m = int(input())
result = []

graph = [[] for _ in range(n+1)]
visited = [0] * (n+1)

for i in range(m):
    a, b = map(int, sys.stdin.readline().split())
    graph[a].append(b)
    graph[b].append(a)

# for i in range(n+1):
#     graph[i].sort()

dfs(x, 0)

if len(result) == 0:
    print(-1)
else:
    print(result[0]-1)

File: 백준/Silver/2644. 촌수계산/촌수계산.java
import java.util.*;
import java.io.*;

public class Main {
    static final int max = 100+10;
    static int n;
    static int start, end;
    static int num;
    static int answer = -1;

    static boolean visited[];
    static boolean graph[][];

    public static void dfs(int idx, int count){
        visited[idx] = true;
        if (idx == end){
            answer = count;
            return;
        }

        for (int i=1; i<=n; i++){
            if (visited[i] == false && graph[idx][i] == true) {
                dfs(i, count+1);
            }
        }
    }
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        n = Integer.parseInt(br.readLine());

        StringTokenizer st = new StringTokenizer(br.readLine());

        start = Integer.parseInt(st.nextToken());
        end = Integer.parseInt(st.nextToken());
        num = Integer.parseInt(br.readLine());

        graph = new boolean[max][max];
        visited = new boolean[max];


        for (int i=0; i<num; i++){
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            graph[x][y] = true;
            graph[y][x] = true;
        }

        dfs(start, 0);

        bw.write(String.valueOf(answer));

        br.close();
        bw.close();
    }
}
